src/backend/commands/analyze.c

```diff
@@ -2625,66 +2549,93 @@ acquire_sample_rows_dispatcher(Relation onerel, bool inh, int elevel,
 	 * Read the result set from each segment. Gather the sample rows *rows,
 	 * and sum up the summary rows for grand 'totalrows' and 'totaldeadrows'.
 	 */
-	funcRetValues = (char **) palloc0(funcTupleDesc->natts * sizeof(char *));
+	funcRetValues = (Datum *) palloc0(funcTupleDesc->natts * sizeof(Datum));
 	funcRetNulls = (bool *) palloc(funcTupleDesc->natts * sizeof(bool));
-	values = (char **) palloc0(relDesc->natts * sizeof(char *));
+	dvalues = (Datum *) palloc0(relDesc->natts * sizeof(Datum));
+	dnulls = (bool *) palloc(relDesc->natts * sizeof(bool));
 	sampleTuples = 0;
 	*totalrows = 0;
 	*totaldeadrows = 0;
-	for (int resultno = 0; resultno < cdb_pgresults.numResults; resultno++)
+
+	slot = MakeSingleTupleTableSlot(queryDesc->tupDesc);
+	while (tuplestore_gettupleslot(portal->holdStore, true, false, slot))
 	{
-		struct pg_result *pgresult = cdb_pgresults.pg_results[resultno];
-		bool		got_summary = false;
+		TupleDesc	typeinfo = slot->tts_tupleDescriptor;
+		int			natts = typeinfo->natts;
+		Datum		attr;
+		bool		isnull;
 		double		this_totalrows = 0;
 		double		this_totaldeadrows = 0;
 
-		if (PQresultStatus(pgresult) != PGRES_TUPLES_OK)
+		/* There should be only one attribute with OID RECORDOID. */
+		if (1 != natts)
 		{
-			cdbdisp_clearCdbPgResults(&cdb_pgresults);
-			ereport(ERROR,
-					(errmsg("unexpected result from segment: %d",
-							PQresultStatus(pgresult))));
+			elog(ERROR,
+				 "wrong number of attributes %d when 1 expected",
+				 natts);
 		}
+		if (RECORDOID != typeinfo->attrs[0]->atttypid)
+		{
+			elog(ERROR,
+				 "wrong attribute OID %d, RECORDOID %d is expected",
+				 typeinfo->attrs[0]->atttypid, RECORDOID);
 
-		if (GpPolicyIsReplicated(onerel->rd_cdbpolicy))
+		}
+
+		/* Make sure the tuple is fully deconstructed */
+		slot_getallattrs(slot);
+
+		/* There should be only one attribute with OID RECORDOID */
+		attr = slot_getattr(slot, 1, &isnull);
```

Comment:
1. We extract a single record's datum from a memory tuple.
2. Then we use this datum to construct a new heap tuple.
3. We detoast this new heap tuple with a real tuple description.

So my question is - how do we serialize this datum as we can detoast it only with heap tuple?

Reply:
Analyze uses [gp_acquire_sample_rows](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/commands/analyzefuncs.c#L102) function to acquire a sample set of rows from table with an additional information (totalrows, totaldeadrows and oversized_cols_bitmap) as summary row. Sample rows are returned first, then summary row.
This function constructs [heap tuple from result values](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/commands/analyzefuncs.c#L321) and returns the [datum from it](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/commands/analyzefuncs.c#L325)), [the same for the summary row](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/commands/analyzefuncs.c#L345-L349).